openpype/plugins/publish/integrate_inputlinks_ayon.py

```diff
@@ -0,0 +1,160 @@
+        """
+
+        if not AYON_SERVER_ENABLED:
+            self.log.info("Skipping, not in AYON mode")
+            return
+
+        workfile_instance, other_instances = self.split_instances(context)
+
+        # Variable where links are stored in submethods
+        new_links_by_type = collections.defaultdict(list)
+
+        self.create_workfile_links(
+            workfile_instance, other_instances, new_links_by_type)
+
+        self.create_generative_links(other_instances, new_links_by_type)
+
+        self.create_links_on_server(context, new_links_by_type)
+
+    def split_instances(self, context):
+        workfile_instance = None
+        other_instances = []
+
+        for instance in context:
+            # Skip inactive instances
+            if not instance.data.get("publish", True):
+                continue
+
+            version_doc = instance.data.get("versionEntity")
+            if not version_doc:
+                self.log.debug(
+                    "Instance {} doesn't have version.".format(instance))
+                continue
+
+            family = instance.data.get("family")
+            if family == "workfile":
+                workfile_instance = instance
+            else:
+                other_instances.append(instance)
+        return workfile_instance, other_instances
+
+    def add_link(self, new_links_by_type, link_type, input_id, output_id):
+        """Add dependency link data into temporary variable.
+
+        Args:
+            new_links_by_type (dict[str, list[dict[str, Any]]]): Object where
+                output is stored.
+            link_type (str): Type of link, one of 'reference' or 'generative'
+            input_id (str): Input version id.
+            output_id (str): Output version id.
+        """
+
+        new_links_by_type[link_type].append((input_id, output_id))
+
+    def create_workfile_links(
+        self, workfile_instance, other_instances, new_links_by_type
+    ):
+        if workfile_instance is None:
+            self.log.warn("No workfile in this publish session.")
+            return
+
+        workfile_version_id = workfile_instance.data["versionEntity"]["_id"]
+        # link workfile to all publishing versions
+        for instance in other_instances:
+            self.add_link(
+                new_links_by_type,
+                "generative",
+                workfile_version_id,
+                instance.data["versionEntity"]["_id"],
+            )
+
+        loaded_versions = workfile_instance.context.get("loadedVersions")
+        if not loaded_versions:
+            return
+
+        # link all loaded versions in scene into workfile
+        for version in loaded_versions:
+            self.add_link(
+                new_links_by_type,
+                "reference",
+                version["version"],
+                workfile_version_id,
+            )
+
+    def create_generative_links(self, other_instances, new_links_by_type):
```

Comment:
I'd say that this will be most of the time `reference` link. It's actually not 100% the case, but if a version is loaded and is tracked in the inventory, it is mose likely being pulled from the disk into current workfile, hence `reference`

`generative` is relationship between scene and it's render for example, where you don't need the scene to use the render once it's generated

Reply:
To be honest I just "copied" code from previous integrator. If `"inputVersions"` represents loaded versions then what is different in that and these [here](https://github.com/ynput/OpenPype/pull/4842/files#diff-954c39e9258e836148aec91c00df5694700889aea4c2d1333bd09e4bc39bc308R86) ?.

Reply:
Ok so input versions is more specific links based on scene. That is available only in fusion, houdini and maya.

Reply:
Yes, input versions are what actually ends up being used for the instance - say I have loaded a cube and sphere in my scene but the pointcacheMain publish from my scene only contains the cube, then it doesn't have the sphere as input version.

The workfile publish however, would always have both. As such if you still want to know what was in the scene at the time you can go:
- `pointcacheMain` -> directly included content
- `pointcacheMain` -> `workfile` -> all contents

The workfile is always added into the published subset and thus is always accessible like this.

Reply:
Just a not for future this is actualy.

pointcacheMain **-gen->** directly included content
pointcacheMain **-ref->** workfile **-gen->** all contents